pingcap/tidb-operator
# support passing raw toml config for tiflash

### What problem does this PR solve? 

support passing raw toml config for tiflash

follow up #3327

### What is changed and how does it work?

origin way to set config:
```yaml
    tiflash:
      baseImage: pingcap/tiflash
      maxFailoverCount: 3
      replicas: 1
      storageClaims:
        - resources:
            requests:
              storage: 1Gi
--        # storageClassName: local-storage
      # requests:
      #   storage: "1Gi"
      config:
        config:
          logger:
            level: "warn"
            count: 9
        proxy:
          log-level: "warn"
          gc:
            batch-keys: 501
```
after this pr, support pass config like this:
```yaml
    tiflash:
      baseImage: pingcap/tiflash
      maxFailoverCount: 3
      replicas: 1
      storageClaims:
        - resources:
            requests:
              storage: 1Gi
--        # storageClassName: local-storage
      # requests:
      #   storage: "1Gi"
      config:
        config: |
          [logger]
            level = "warn"
            count = 9
        proxy: |
          log-level = "warn"
          [gc]
            batch-keys = 501
```

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Deploy cluster using the flowing yaml(one use raw, one use origin way),
Check pos normal and config in configmaps as expect

<details><summary>tidb-cluster.yaml CLICK ME</summary>
<p>


```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: basic
spec:
  version: v4.0.6
  timezone: UTC
  pvReclaimPolicy: Delete
  discovery: {}
  configUpdateStrategy: RollingUpdate
  pd:
    baseImage: pingcap/pd
    replicas: 1
    # if storageClassName is not set, the default Storage Class of the Kubernetes cluster will be used
    # storageClassName: local-storage
    requests:
      storage: "1Gi"
    config: |
      [schedule]
        max-merge-region-keys = 22222
        max-merge-region-size = 22
  tikv:
    baseImage: pingcap/tikv
    replicas: 1
    # if storageClassName is not set, the default Storage Class of the Kubernetes cluster will be used
    # storageClassName: local-storage
    requests:
      storage: "1Gi"
    config: |
      [storage]
        # In basic examples, we set this to avoid using too much storage.
        reserve-space = "2MB"
  tidb:
    baseImage: pingcap/tidb
    replicas: 1
    service:
      type: ClusterIP
    config:
      token-limit: 203
      max-server-connections: 51
  pump:
    baseImage: pingcap/tidb-binlog
    replicas: 1
    # storageClassName: local-storage
    requests:
      storage: 20Gi
    # schedulerName: default-scheduler
    # config:
    #   gc: 7
    #   storage:
    #     kv:
    #       compaction-total-size-multiplier: 8.0
    config: |
      gc = 6
      [storage]
        [storage.kv]
          compaction-total-size-multiplier = 8.0


  tiflash:
    baseImage: pingcap/tiflash
    maxFailoverCount: 3
    replicas: 1
    storageClaims:
      - resources:
          requests:
            storage: 1Gi
        # storageClassName: local-storage
    # requests:
    #   storage: "1Gi"
    config:
      config: |
        [logger]
          level = "warn"
          count = 9
      proxy: |
        log-level = "warn"
        [gc]
          batch-keys = 501
```

</p>
</details>

<details><summary>tidb-cluster-ym.yamlCLICK ME</summary>
<p>


```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: basic-ym
spec:
  version: v4.0.6
  timezone: UTC
  pvReclaimPolicy: Delete
  discovery: {}
  configUpdateStrategy: RollingUpdate
  pd:
    baseImage: pingcap/pd
    replicas: 1
    # if storageClassName is not set, the default Storage Class of the Kubernetes cluster will be used
    # storageClassName: local-storage
    requests:
      storage: "1Gi"
    config:
      schedule:
        max-merge-region-keys: 22222
        max-merge-region-size: 22
  tikv:
    baseImage: pingcap/tikv
    replicas: 1
    # if storageClassName is not set, the default Storage Class of the Kubernetes cluster will be used
    # storageClassName: local-storage
    requests:
      storage: "1Gi"
    config:
      storage:
        # In basic examples, we set this to avoid using too much storage.
        reserve-space: "2MB"
  tidb:
    baseImage: pingcap/tidb
    replicas: 1
    service:
      type: ClusterIP
    config:
      token-limit: 203
      max-server-connections: 51
  pump:
    baseImage: pingcap/tidb-binlog
    replicas: 1
    # storageClassName: local-storage
    requests:
      storage: 20Gi
    # schedulerName: default-scheduler
    config:
      gc: 6

  tiflash:
    baseImage: pingcap/tiflash
    maxFailoverCount: 3
    replicas: 1
    storageClaims:
      - resources:
          requests:
            storage: 1Gi
        # storageClassName: local-storage
    # requests:
    #   storage: "1Gi"
    config:
      config:
        logger:
          level: "warn"
          count: 9
      proxy:
        log-level: "warn"
        gc:
          batch-keys: 501
```

</p>
</details>


Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
support passing raw toml config for tiflash
```
